[1단계 - 지하철 정보 관리 기능] 필립(양재필) 미션제출합니다.

pilyang opened the pull request:
안녕하세요 럿고!

1단계 미션 제출합니다.

처음에는 domain 위주로 생각을 해보기 위해서 Repository는 db연결을 하지 않은 SimpleRepsitory를 이용하며 TDD사이클을 지키며 하려 노력했습니다. 하지만 후에 시간이 부족하여 tdd를 잘 진행하지 못하여 테스트가 부족한 상황입니다. 해당 부분들은 리팩터링을 하며 가능하면 보완할 수 있도록 하겠습니다.

사용한 SimpleRepository는 DbRepository구현후 제거하였습니다.

---

지하철 노선의 저장 방식으로는 `Line`도메인이 상행(up station)부터 하행(down station)방향으로 순서대로 StationEdge리스트를 가지도록 설계하였습니다.

`Station Edge`의 구조는 아래와 같이 구상하고 구현하였습니다.
```
O: station (id)
--: station 간 distance
(상행방향) --O  --O  --O (하행방향)
```

하나의 edge가 하나의 역과 상행방향의 인접한 역까지의 거리를 가지고 있도록 구현했습니다. 이때 상행 종점역은 distance가 0을 가지고 있도록 했습니다.

---

시간안에 API기능요구사항을 만족시키기 위해서 해당 요구사항을 위해 당장 필요한것이 아니라고 생각한 api들은 아직 구현하지 않았고, 관련 테스트들도 disabled처리를 해두었습니다.

도메인 분리등 부족한부분이 많은 코드입니다...

부족하고, 놓치고 있는 부분들 따끔한 리뷰 해주시면 리팩터링 하면서 고민해보고 반영해볼 수 있도록 하겠습니다.

잘부탁드립니다.
감사합니다!

kyY00n committed:
docs: API 기능 요구사항 추가

kyY00n committed:
test: 역 관련 API 테스트 수정

kyY00n committed:
test: 노선 관련 API 테스트 수정

kyY00n committed:
test: 노선 관련 API 테스트 수정

kyY00n committed:
docs: 비즈니스 요구사항 작성

kyY00n committed:
feat: 역 등록 기능 구현

kyY00n committed:
style: reformat code

kyY00n committed:
refactor: 협력객체를 실제 객체로 사용

kyY00n committed:
feat: 노선 생성 기능

kyY00n committed:
feat: 노선 중복이름 검증 구현

kyY00n committed:
feat: 노선에 역을 등록하는 기능 구현

kyY00n committed:
feat: 노선에서 역을 제거하는 기능

* 역이 3개 이상인 노선

kyY00n committed:
feat: 역이 2개인 노선에서 역을 제거하는 기능

kyY00n committed:
feat: 구현된 기능 컨트롤러 연결

kyY00n committed:
feat: 노선 조회 API 구현

kyY00n committed:
feat: 전체 노선 조회 API 구현

kyY00n committed:
feat: db repository구현 및 Respository 교체

kyY00n committed:
style: 코드 reformat, 사용하지 않는 import 제거

kyY00n committed:
refactor: 역 추가 메서드 분리

kyY00n committed:
refactor: Line 생성 로직변경

ksy90101 commented on src/main/java/subway/dao/LineDao.java:
```diff
@@ -1,61 +1,69 @@
 
-@Repository
+@Component
 public class LineDao {
```
Dao를 Interface로 빼서 공통 추상화를 해보면 어떨까요?

pilyang replied:
`Dao`라는 interface를 만들어서 기본적인 crud (`insert`, `findById`, `findAll`, `update`, `delete`)를 만들 수 있도록 해라로 이해했는데, 제가 이해한 것이 맞나요?

해당 방식으로 리팩터링을 해보니 기본적인 기능은 다 만들어야하도록 강제해서 놓치지 않는다는 점과, 제 dao의 경우 공통 코드부을 default method로 만들어서 사용할 수 있다정도의 장점은 느낄 수 있었습니다.

그런데 이 외에 각각의 데이터 테이블마다 따로 필요한 메서드도 만들어줘야하고, 필수적으로 강제하여 만든 메서드들을 사용하지 않는 경우도 있었습니다. 그래서 꼭 이렇게 공통추상화를 했어야 했나 싶은 의문이 생깁니다.

혹시 제가 생각하지 못하거자, 잘못 적용한 부분이 있으면 럿고의 의견을 들려주실 수 있으신가요?

ksy90101 replied:
맞습니다.
사용하지 않는 경우엔 Exception을 던지는 로직을 작성해도 될거 같습니다.
차후에 JDBC가 아니라 JPA나 이런걸 사용하게 되었을때 쉽게 갈아끼우기 쉽게 하는게 목적이긴 했어요!

ksy90101 commented on src/main/java/subway/domain/Line.java:
```diff
@@ -1,24 +1,123 @@
+            long adjacentStationId,
+            int distance
+    ) {
+        if (contains(insertStationId)) {
+            throw new StationAlreadyExistsException();
+        }
```
이런 validate로직이 중복되기 떄문에 메서드로 제거해서 진행하면 어떨까요?

pilyang replied:
검증로직 메소드 분리 했습니다!

ksy90101 commented on src/main/java/subway/domain/Line.java:
```diff
@@ -1,24 +1,123 @@
+        final int adjacentDownStationIndex = stationEdges.indexOf(getStationEdge(adjacentStationId)) + 1;
+
+        boolean isLastEdge = adjacentDownStationIndex == stationEdges.size();
+        if (isLastEdge) {
+            StationEdge insertedStationEdge = new StationEdge(insertStationId, distance);
+            stationEdges.add(insertedStationEdge);
+            return new InsertionResult(insertedStationEdge, null);
+        }
```
LastEdge는 특이한 케이스 같은데, 메서드로 분리해서 따로 로직을 구분해보면 어떨까요?

pilyang replied:
분리가 필요하다 생각되는 부분들 분리했습니다!

ksy90101 commented on src/main/java/subway/domain/Line.java:
```diff
@@ -1,24 +1,123 @@
     private Long id;
     private String name;
     private String color;
+    private List<StationEdge> stationEdges;
```
중복된 station이 되지 않으니, 이런경우 Set을 써서 중복 방지를 해보면 어떨까요?

pilyang replied:
지금 제 로직에서는 stationEdge가 하행방향 역의 정보만 가지고 있고, 몇번째 역인지에대한 순서 또한 가지고 있지 않아서 List의 순서로 노선도의 순서를 가지게 했습니다. 이러한 이유로 Set이 아닌 List를 통해 순서를 보장하도록 했습니다.

하지만 이런식으로 구현을 하니 럿고께서 아래 line repository 코멘트에서 남겨주신것과 같이 레포지토리에서도 너무 도메인에 로직에 대한 구현이 있는 느낌이 들고, 도메인 또한 레포지토리에서 가져온 순서가 맞다는 확신을 가지고 가야하기에 불안한 부분이 있는 것 같습니다.

이부분은 그래서 2단계 기능구현 전에 StationEdge가 상행과 하행의 정보를 가지고 있도록 하여 레포지토리에서 순서를 보장해준다는 전제 없이도 도메인만으로 신뢰할 수 있도록 리팩터링 해 볼 계획입니다. 이러한 방식으로 리팩터링을 하면 말씀해주신대로 Set을 활용하여 중복방지도 될 것 같습니다.

ksy90101 replied:
네. 순서는 order by로 하면 될거 같고.
LinkedHashSet을 사용해보면 어떨까 합니다.

ksy90101 commented on src/main/java/subway/domain/Line.java:
```diff
@@ -1,24 +1,123 @@
+    private StationEdge getStationEdge(long stationId) {
+        return stationEdges.stream()
+                .filter(edge -> edge.getDownStationId().equals(stationId))
+                .findFirst().orElseThrow(StationNotFoundException::new);
```
```suggestion
                .findFirst()
                .orElseThrow(StationNotFoundException::new);
```

ksy90101 commented on src/main/java/subway/domain/Line.java:
```diff
@@ -1,24 +1,123 @@
+
+    private StationEdge getStationEdge(long stationId) {
+        return stationEdges.stream()
+                .filter(edge -> edge.getDownStationId().equals(stationId))
```
이 메서드도 edge에게 메시지를 물어보면 깔끔해 지겠네요~

pilyang replied:
level1에서 신경쓰고 공부하던 부분인데 놓치게 되었었네요ㅠㅜ
감사합니다!!

ksy90101 commented on src/main/java/subway/service/LineService.java:
```diff
@@ -0,0 +1,99 @@
+        LineDirection direction = LineDirection.valueOf(stationInsertRequest.getDirection());
+        int distance = stationInsertRequest.getDistance();
+
+        if (direction == LineDirection.UP) {
+            return line.insertUpStation(stationId, adjacentStationId, distance);
+        }
+        return line.insertDownStation(stationId, adjacentStationId, distance);
```
insertStation을 하면 Up, Dwon인지는 line이 알아서 처리하게 하면 어떨까요?
굳이 서비스에서 이렇게 분기를 칠 필요가 있을까 해서요.

pilyang replied:
네 생각해보니 도메인이 해당 역할을 가지고 있는게 맞을것 같다고 생각을 해서 수정했습니다!

ksy90101 commented on src/main/java/subway/service/LineService.java:
```diff
@@ -0,0 +1,99 @@
+        this.stationRepository = stationRepository;
+    }
+
+    public Long create(LineRequest lineRequest) {
```
전체적으로 Transaction이 없네요 :)

pilyang replied:
헉... 없는부분들 추가 했습니다!!

ksy90101 commented on src/main/java/subway/repository/DbLineRepository.java:
```diff
@@ -0,0 +1,157 @@
+
+
+@Repository
+public class DbLineRepository implements LineRepository {
```
Repository의 로직들에서 서비스 로직인것도 많은거 같아요.
Repository나 Dao는 단순히 DB와 통신하는 역할을 하는건데, 실제 도메인들끼리의 로직까지 역할을 하는거 같은데, 혹시 이런 구조를 구현한 이유가 있을까요?

pilyang replied:
지금 Repository내부에 있는 로직들은 비지니스 요구사항, 로직들을 위한 것이 아닌 table내부 데이터인 entity에서 domain으로 변환하고, 변환을 위한 정렬 로직들이 포함되어있습니다.
이부분은 db에서 도메인으로의 변환 로직이라 repository에 구현하게 되었습니다.

변환과 정렬 과정도 domain에 있게 되면 domain이 특정 db에 종속되는 entity를 알고, db태이블의 값을 이용해서 정렬을 하는 것 같아서 domain에서 구현을 안하기도 하였습니다.

혹시 이러한 부분들도 domain이 할 수 있도록 하는편이 더 좋았을까요..?

다만 이부분은 앞서 StationEdge를 순서를 보장하는 list로 가지고 있는 구조라서 변환과 정렬 로직이 더 필요했던것 같습니다. 만약에 해당 부분을 리팩터링 하게되면 해당 로직부분이 줄어들거나 이동할 가능성은 있을 것 같습니다.

pilyang replied:
리펙터링을 하면서 해당 부분에대해 추가로 고민해보니, 노선에 역을 추가하고, 지우는 경우 repository의 여러 메서드들을 통해 업데이트 하는 부분을 `insertStationEdge`, `updateStationEdge`, `deleteStation` 3가지 메서드를 조합해서 적용하는 등의 행동이 레포지토리가 너무 도메인에 종속되어 기능을 가진다는 느낌이 들었습니다.

그래서 이 부분들은 `updateStationEdges`메서드 하나만을 두어 line domain에서 station edge들에 변화가 생기면 station edge를 업데이트 해주는 방식으로 리팩터링 하였습니다.

ksy90101 commented on src/main/java/subway/ui/ApiControllerAdvice.java:
```diff
@@ -0,0 +1,36 @@
+    })
+    public ResponseEntity<Void> handleNotFound() {
+        return ResponseEntity.notFound().build();
+    }
+
+
```
```suggestion
    }
```

ksy90101 commented on src/main/java/subway/dto/StationInsertRequest.java:
```diff
@@ -0,0 +1,39 @@
+package subway.dto;
+
+public class StationInsertRequest {
```
Dto에 전체적으로 Validate가 없네요.
기본적인 Validate를 한번 적용하면 어떨까요?

pilyang replied:
Request Validation과 예외상황시 예외메시지 반환 추가했습니다!

ksy90101 commented on src/main/java/subway/repository/DbLineRepository.java:
```diff
@@ -0,0 +1,157 @@
+        StationEdgeEntity first = stationEdgeEntities.stream()
+                .filter(stationEdgeEntity -> stationEdgeEntity.getPreviousStationEdgeId() == null)
+                .findFirst()
+                .get();
```
get을 쓰면 null과 다를바 없기 때문에, get보단 NPE 방어를 해보시면 좋을거 같습니다.

pilyang replied:
앗.. 관련 예외를 만들어두고 그냥 get을 사용했었네요...
없으면 예외를 발생시키도록 변경했습니다!!

ksy90101 commented on src/main/java/subway/repository/DbLineRepository.java:
```diff
@@ -0,0 +1,157 @@
+
+    private static StationEdgeEntity getNext(List<StationEdgeEntity> stationEdgeEntities,
+                                             Long previousId) {
+        StationEdgeEntity next = stationEdgeEntities.stream()
+                .filter(stationEdge -> stationEdge.getPreviousStationEdgeId().equals(previousId))
+                .findFirst().get();
+        return next;
```
별건 아니지만, ealry retrun을 해보면 어떨까 해요.
여기도 get을 그냥 쓰네요 ㅠㅠ
```suggestion
        return stationEdgeEntities.stream()
                .filter(stationEdge -> stationEdge.getPreviousStationEdgeId().equals(previousId))
                .findFirst().get();```

ksy90101 commented on src/test/java/subway/integration/LineIntegrationTest.java:
```diff
@@ -138,14 +164,15 @@ void getLine() {
         assertThat(resultResponse.getId()).isEqualTo(lineId);
     }
 
+    @Disabled
```
Disabled가 있는 이유가 있나요?

pilyang replied:
이번 미션에서 기본으로 제공해준 인수테스트들 입니다. 구현을 하면서 이번 1단계에서는 필요가 없다고 판단을 해 구현을 하지 않을 api테스트들 인데, 이후 단계를 진행시 필요한 기능들이 있을 수 있다고 판단하여 완전히 지우기보다는 우선은 disabled로 표시해두었습니다.
이후 단계를 진행하면서 필요한 부분들은 그대로 또는 수정을 하여 사용을 하고, 전체적인 기능구현에 필요가 없는 api라고 판단된 부분들은 완전히 제거를 할 계획입니다.

ksy90101 reviewed (changes requested):
안녕하세요 :)
지하철 미션하시느라고 고생많으셨어요~
간단한 피드백을 남겼으니 1단계에서 조금 수정하고 가면 좋을거 같습니다.
질문이 있다면 언제든 슬랙으로 남겨주세요~

pilyang committed:
style: final 추가, 코드 컨벤션 맞추기

pilyang committed:
refactor: Dao interface를 통해 공통 추상화

- Dao interface 적용
- Entity interface 적용

pilyang committed:
refactor: Line, StationEdge 메소드 분리

pilyang committed:
feat: Request Body검증 추가

- Controller Dto package 위치 변경
- RequestBody 검증 추가
- 예외반환시 예외내용 예외 response body에 추가

pilyang committed:
refactor: stationEdge의 변경사항 적용메서드 update로 통일

pilyang committed:
fix: transactional annotation 추가

pilyang committed:
refactor: 역 추가 방향확인에 대한 책임 Line으로 이동

pilyang reviewed (commented):
안녕하세요 럿고!

1차 리뷰 반영해서 수정하였습니다!

코멘트 답글로도 남긴, StationEdge의 구조 등 생각해본 domain관련 부분에서의 리팩터링까지 모두 하고 보내기에는 시간이 오래 걸릴것 같아서 우선 리뷰 남겨주신 부분들 먼저 수정을 완료했고, 리뷰요청 보냅니다.
해당 부분의 리펙터링은 리뷰요청을 보낸 후 진행하며 2단계 혹은 다음 리뷰요청에서 반영하도록 하겠습니다.

감사합니다!

pilyang reviewed (commented):
안녕하세요 럿고!

1차 리뷰 반영해서 수정하였습니다!

코멘트 답글로도 남긴, StationEdge의 구조 등 생각해본 domain관련 부분에서의 리팩터링까지 모두 하고 보내기에는 시간이 오래 걸릴것 같아서 우선 리뷰 남겨주신 부분들 먼저 수정을 완료했고, 리뷰요청 보냅니다.
해당 부분의 리펙터링은 리뷰요청을 보낸 후 진행하며 2단계 혹은 다음 리뷰요청에서 반영하도록 하겠습니다.

감사합니다!

ksy90101 reviewed (approved):
필립 안녕하세요 :)
피드백 적용하시느라 고생 하셧습니다~
전체적으로 잘 된거 같아서 이만 마무리할게요~
고생 많으셨어요! 2단계에서 봐요!

ksy90101 commented on src/main/java/subway/service/LineService.java:
```diff
@@ -0,0 +1,96 @@
+
+    @Transactional
+    public void insertStation(final StationInsertRequest stationInsertRequest) {
+        findStationById(stationInsertRequest.getStationId());
+        findStationById(stationInsertRequest.getAdjacentStationId());
```
쿼리는 최소화 하는게 좋으니깐요.
이런 경우 ids를 사용해서 한개의 쿼리만 날라가게 하는것도 고려해볼수 있을거 같습니다.
